core/lib/state/src/rocksdb/mod.rs

```diff
@@ -131,23 +199,162 @@ impl RocksdbStorage {
         tracing::info!(
             "Secondary storage for L1 batch #{latest_l1_batch_number} initialized, size is {estimated_size}"
         );
+
+        self.save_missing_enum_indices(conn).await;
+    }
+
+    async fn apply_storage_logs(
+        &mut self,
+        storage_logs: HashMap<StorageKey, H256>,
+        conn: &mut StorageProcessor<'_>,
+    ) {
+        let (logs_with_known_indices, logs_with_unknown_indices): (Vec<_>, Vec<_>) = self
+            .process_transaction_logs(storage_logs)
+            .into_iter()
+            .partition_map(|(key, StateValue { value, enum_index })| match enum_index {
+                Some(index) => Either::Left((key, (value, index))),
+                None => Either::Right((key, value)),
+            });
+        let keys_with_unknown_indices: Vec<_> = logs_with_unknown_indices
+            .iter()
+            .map(|(key, _)| key.hashed_key())
+            .collect();
+
+        let enum_indices = conn
+            .storage_logs_dedup_dal()
+            .enum_indices_for_keys(&keys_with_unknown_indices)
+            .await;
+        assert_eq!(keys_with_unknown_indices.len(), enum_indices.len());
+        self.pending_patch.state = logs_with_known_indices
+            .into_iter()
+            .chain(
+                logs_with_unknown_indices
+                    .into_iter()
+                    .map(|(key, value)| (key, (value, enum_indices[&key.hashed_key()]))),
+            )
+            .collect();
+    }
+
+    async fn save_missing_enum_indices(&self, conn: &mut StorageProcessor<'_>) {
+        let (Some(start_from), true) = (
+            self.enum_migration_start_from(),
+            self.enum_index_migration_chunk_size > 0,
+        ) else {
+            return;
+        };
+
+        let started_at = Instant::now();
+        tracing::info!(
+            "RocksDB enum index migration is not finished, starting from key {start_from:0>64x}"
+        );
+
+        let mut write_batch = self.db.new_write_batch();
+        let (keys, values): (Vec<_>, Vec<_>) = self
+            .db
+            .from_iterator_cf(StateKeeperColumnFamily::State, start_from.as_bytes())
+            .filter_map(|(key, value)| {
+                (!Self::is_special_key(&key) && value.len() != 40)
+                    .then(|| (H256::from_slice(&key), H256::from_slice(&value[..32])))
+            })
+            .take(self.enum_index_migration_chunk_size)
+            .unzip();
+        let enum_indices = conn
+            .storage_logs_dedup_dal()
+            .enum_indices_for_keys(&keys)
+            .await;
+        assert_eq!(keys.len(), enum_indices.len());
+
+        for (key, value) in keys.iter().zip(values) {
+            let index = enum_indices[key];
+            write_batch.put_cf(
+                StateKeeperColumnFamily::State,
+                key.as_bytes(),
+                &StateValue {
+                    value,
+                    enum_index: Some(index),
+                }
```

Comment:
Nit: Having a constructor for `StateValue` could improve readability in such multi-line statements: `StateValue::new(value, index).serialize()` is arguably easier to read.